src/main/java/com/android/volley/RequestQueue.java

```diff
@@ -38,11 +38,29 @@
     // TODO: This should not be a generic class, because the request type can't be determined at
     // compile time, so all calls to onRequestFinished are unsafe. However, changing this would be
     // an API-breaking change. See also: https://github.com/google/volley/pull/109
+    @Deprecated // Use RequestEventListener instead.
     public interface RequestFinishedListener<T> {
         /** Called when a request has finished processing. */
         void onRequestFinished(Request<T> request);
     }
 
+    /** Request event types listeners will be notified about. */
+    public enum RequestEvent {
+        REQUEST_QUEUED,
+        REQUEST_CACHE_LOOKUP_STARTED,
+        REQUEST_CACHE_LOOKUP_FINISHED,
+        REQUEST_PROCESSING_STARTED,
+        REQUEST_PROCESSING_FINISHED,
+        REQUEST_FINISHED
+    }
+
+    /** Callback interface for request life cycle events. */
+    public interface RequestEventListener {
+        // Listener can be called from different threads but in synchronized way. Also it's
```

Comment:
I feel like you should call out that any event processing should be kept to a minimum or moved to a separate thread to prevent the request from blocking. Alternatively, require an Executor be provided with the listener and just post all of the events to it.

Reply:
Done